File: Leetcode/Array/016_3Sum_Closest.java
public class Solution {
    // O(n^2) time, O(1) space
    public int threeSumClosest(int[] nums, int target) {
        if(nums == null || nums.length == 0)
            return 0;
        
        Arrays.sort(nums);
        int ret = Integer.MAX_VALUE;
        int diff = Integer.MAX_VALUE;
        
        for(int i = 0; i < nums.length - 2; i++){
            int l = i + 1, r = nums.length - 1;
            while(l < r){
                int sum = nums[i] + nums[l] + nums[r];
                int curDiff = Math.abs(target - sum);
                if(curDiff < diff){
                    diff = curDiff;
                    ret = sum;
                }
                if(sum > target)
                    r--;
                else
                    l++;
            }
            
        }
        return ret;
    }
}

File: Leetcode/All/016_3Sum_Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int ret = nums[0] + nums[1] + nums[2];
        int diff = abs(target - ret);
        for(int i = 0; i < nums.size() - 2; i++){
            int l = i + 1, r = nums.size() - 1;
            while(l < r){
                int sum = nums[i] + nums[l] + nums[r];
                if(sum == target)   
                    return target;
                int new_diff = abs(sum - target);
                if(new_diff < diff){
                    diff = new_diff;
                    ret = sum;
                }
                if(sum > target){
                    r--;
                }else if(sum < target){
                    l++;
                }
            }
        }
        return ret;
    }
};